### PR TITLE
Add TypeDescriptionProviders to XLinq

### LINKED ISSUE #32641: WPF on .NET Core needs TypeDescriptor on System.Xml.Linq.XElement
WPF has extension libraries for various types.  One of them is for System.Xml.Linq.XElement.

It expects that the TypeDescriptor for XElement contains the following property descriptors:
- Elements
- Descendants
- Attribute
- Element

This was previously provided by [XTypeDescriptor<T>.GetProperties with a special case for XElement](https://referencesource.microsoft.com/#System/compmod/system/componentmodel/CustomTypeDescriptor.cs,3575006103cb88da).  These seem to not exist in .NET Core and neither does the association of the TypeDescriptor with XElement ([see here](https://referencesource.microsoft.com/#System.Xml.Linq/System/Xml/Linq/XLinq.cs,3367036406d1344a)).

WPF requires this in SystemXmlLinqExtension (the implementation of this is not on the external reference source so I cannot link it here).

cc: @vatsan-madhavan, @ericstj

### PR DESCRIPTION
**Pull request by ericstj**
This adds TypeDescriptionProviders to XAttribute and XElement

To do this without introducing a TypeConverter dependency in XML I had to push
TypeDescriptionProviderAttribute down.

This still causes TypeConverter to gain a dependency on XML, but right now that's the direction
we've been heading with TypeConverter.

I had to use the string overload of TypeDescriptionProviderAttribute in order to have a soft
dependency on TypeConverter.

Fixes #32641

### CODE REVIEW

File: src/System.ComponentModel.TypeConverter/src/MS/Internal/Xml/Linq/ComponentModel/XComponentModel.cs [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Xml.Linq;
+
+namespace MS.Internal.Xml.Linq.ComponentModel
+{
+    class XTypeDescriptionProvider<T> : TypeDescriptionProvider
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/MS/Internal/Xml/Linq/ComponentModel/XComponentModel.cs]: Nit: missing explicit visiblity here and elsewhere, e.g. internal

File: src/System.ComponentModel.TypeConverter/src/MS/Internal/Xml/Linq/ComponentModel/XComponentModel.cs [hunk 2]
```diff
@@ -0,0 +1,576 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/MS/Internal/Xml/Linq/ComponentModel/XComponentModel.cs]: I assume all of this is just ported effectively as-is from netfx?

File: src/System.ComponentModel.TypeConverter/src/MS/Internal/Xml/Linq/ComponentModel/XComponentModel.cs [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+                    changeState = a != null && value.element == a.Parent && value.name == a.Name ? a : null;
+                    break;
+            }
+        }
+    }
+
+    class XElementDescendantsPropertyDescriptor : XPropertyDescriptor<XElement, IEnumerable<XElement>>
+    {
+        XDeferredAxis<XElement> value;
+        XName changeState;
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/MS/Internal/Xml/Linq/ComponentModel/XComponentModel.cs]: Nit: missing _ prefix on fields

File: src/System.ComponentModel.TypeConverter/src/MS/Internal/Xml/Linq/ComponentModel/XComponentModel.cs (hunk 2 above)
Comment (ericstj) [reply src/System.ComponentModel.TypeConverter/src/MS/Internal/Xml/Linq/ComponentModel/XComponentModel.cs]: Yes, as is: I kept the dotnet-bot commit to illustrate that.  I'll run the codeformatter to fix up the style issues.  I only ran IDE formatting which missed naming/visibility modifiers.

File: src/System.Xml.XDocument/ref/System.Xml.XDocument.cs [hunk 4]
```diff
@@ -211,6 +212,7 @@ public partial class XDocumentType : System.Xml.Linq.XNode
         public override System.Threading.Tasks.Task WriteToAsync(System.Xml.XmlWriter writer, System.Threading.CancellationToken cancellationToken) { throw null; }
     }
     [System.Xml.Serialization.XmlSchemaProvider(null, IsAny=true)]
+    [System.ComponentModel.TypeDescriptionProvider("MS.Internal.Xml.Linq.ComponentModel.XTypeDescriptionProvider`1[[System.Xml.Linq.XElement, System.Xml.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]],System.ComponentModel.TypeConverter")]
```
Comment (krwq) [inline src/System.Xml.XDocument/ref/System.Xml.XDocument.cs]: on github it looks like there is something wrong with indentation in here (this whole file) - please make sure this is correct

Comment (ericstj) [reply src/System.Xml.XDocument/ref/System.Xml.XDocument.cs]: This file is no longer part of this commit.  I decided not to expose this in the reference assembly.

**Comment by ViktorHofer** [comment]
@ericstj src/MS/Internal/Xml/Linq/ComponentModel/XComponentModel.cs is dead code. Any reason to keep it?

**Comment by ericstj** [comment]
It is not dead code at all.  It's used by the following attributes:
https://github.com/dotnet/corefx/blob/a01c44c78d37267c4e30cb1d647376e103557493/src/System.Private.Xml.Linq/src/System/Xml/Linq/XAttribute.cs#L20
https://github.com/dotnet/corefx/blob/a01c44c78d37267c4e30cb1d647376e103557493/src/System.Private.Xml.Linq/src/System/Xml/Linq/XElement.cs#L35

The typedescriptor tech resolves this dependency at runtime.  Have a look at the tests to see how its used.

**Comment by ViktorHofer** [comment]
I didn't notice that, thanks.

**Pull request merged**